[Refactor] Application Launch Routing

marcomaddy1982 opened the pull request:
The scope of this PR is merging the changes made during the app launch routing refactoring: the main part of the refactoring are: 

[1)](https://github.com/wireapp/wire-ios/pull/4632) Refactor the `AppStateController` and rename it in the `AppStateCalculator`.
[2)](https://github.com/wireapp/wire-ios/pull/4637) Notification Observing Centralisation.
[3)](https://github.com/wireapp/wire-ios/pull/4645)  Decouple the transition animation logic at the launch of the application from the `AppRootViewController` and introduce the `AppRootRouter` and the `RootViewController`.
[4)](https://github.com/wireapp/wire-ios/pull/4648/files) Decouple responsibility from `AppRootRouter`.
[5)](https://github.com/wireapp/wire-ios/pull/4653) Introduced `URLActionRouter` and `SwitchAccountRouter`.
[6)](https://github.com/wireapp/wire-ios/pull/4655) and [7)](https://github.com/wireapp/wire-ios/pull/4656) Clean up and decouple the `AppDelegate`.
[8)](https://github.com/wireapp/wire-ios/pull/4657) Move all the functionality remained in the `AppRootViewController` to the `AppRootRouter`.
[9)](url) DeepLinks Refactoring: Moving all the logic for the deepLinks in the `URLActionRouter` and remove the `whenShowContentDelegateIsAvailable` logic.
[10)](https://github.com/wireapp/wire-ios/pull/4671), [11)](https://github.com/wireapp/wire-ios/pull/4673) and [12)](https://github.com/wireapp/wire-ios/pull/4677) Added tests.
[13)](https://github.com/wireapp/wire-ios/pull/4682) Resolving Conflict from the develop into the feature branch.

### Dependency
sinc-engine: wireapp/wire-ios-sync-engine#1334 (review)

marcomaddy1982 commented:
@billypchan

typfel commented:
Please edit the description to mention that https://github.com/wireapp/wire-ios-sync-engine/pull/1334#pullrequestreview-524006281 is a dependency.

billypchan commented on Wire-iOS Tests/AppStateCalculatorTests.swift:
```diff
@@ -0,0 +1,262 @@
+
+        // THEN
+        XCTAssertEqual(sut.appState, .blacklisted)
+        XCTAssertTrue(delegate.wasNotified)
```
```suggestion
        XCTAssert(delegate.wasNotified)
```
tip: save 4 char to type

typfel replied:
I think `XCTAssertTrue` is better since the intent is more clear. We also use `XCTAssertTrue` almost everywhere in our existing tests.

marcomaddy1982 replied:
Actually I was thinking the same. Probably better return to `XCTAssertTrue`

billypchan commented on Wire-iOS Tests/AppStateCalculatorTests.swift:
```diff
@@ -0,0 +1,262 @@
+        sut.sessionManagerWillLogout(error: error, userSessionCanBeTornDown: nil)
+
+        // THEN
+        XCTAssertEqual(sut.appState, .unauthenticated(error: error as NSError?))
```
the casting is unnecessary?

marcomaddy1982 replied:
Removed

billypchan commented on Wire-iOS/Sources/Main.storyboard:
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="16096" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="XPT-Nt-0JH">
+    <device id="retina6_1" orientation="portrait" appearance="light"/>
+    <dependencies>
+        <deployment identifier="iOS"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="16087"/>
+        <capability name="Safe area layout guides" minToolsVersion="9.0"/>
+        <capability name="documents saved in the Xcode 8 format" minToolsVersion="8.0"/>
+    </dependencies>
+    <scenes>
+        <!--Root View Controller-->
+        <scene sceneID="2L0-l7-fET">
+            <objects>
+                <viewController id="XPT-Nt-0JH" customClass="RootViewController" customModule="Wire" customModuleProvider="target" sceneMemberID="viewController">
+                    <view key="view" contentMode="scaleToFill" id="T9H-gL-yO2">
+                        <rect key="frame" x="0.0" y="0.0" width="414" height="896"/>
+                        <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
+                        <subviews>
+                            <imageView opaque="NO" clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="scaleAspectFit" image="wire-logo-shield" translatesAutoresizingMaskIntoConstraints="NO" id="KML-s6-hr3" userLabel="Image - Logo">
+                                <rect key="frame" x="106" y="317" width="202" height="262"/>
+                            </imageView>
+                        </subviews>
+                        <color key="backgroundColor" white="0.0" alpha="1" colorSpace="custom" customColorSpace="genericGamma22GrayColorSpace"/>
+                        <constraints>
+                            <constraint firstItem="KML-s6-hr3" firstAttribute="centerX" secondItem="T9H-gL-yO2" secondAttribute="centerX" id="8sN-5A-1W8"/>
+                            <constraint firstItem="KML-s6-hr3" firstAttribute="centerY" secondItem="T9H-gL-yO2" secondAttribute="centerY" id="Miv-V9-xEk"/>
+                        </constraints>
+                        <viewLayoutGuide key="safeArea" id="MYb-r6-9JB"/>
+                    </view>
+                </viewController>
+                <placeholder placeholderIdentifier="IBFirstResponder" id="EDv-tZ-FXz" userLabel="First Responder" customClass="UIResponder" sceneMemberID="firstResponder"/>
+            </objects>
+            <point key="canvasLocation" x="-59" y="115"/>
+        </scene>
+    </scenes>
+    <resources>
+        <image name="wire-logo-shield" width="202" height="262"/>
+    </resources>
+</document>
```
Why do we need a storyboard? (for `RootViewController` creation?)

marcomaddy1982 replied:
That one of the idea of the UI Refactoring: introduce storyboards so the ViewController could be more light.

johnxnguyen commented:
@marcomaddy1982 I'm glad to see that code coverage has jumped back up to 60% 👍

marcomaddy1982 commented:
> @marcomaddy1982 I'm glad to see that code coverage has jumped back up to 60% 👍 

Thanks. Let's try to stay on this flow!!!

typfel commented:
Time merge this now or did you want run automation tests one more time?

marcomaddy1982 commented:
> Time merge this now or did you want run automation tests one more time?

@typfel we can decide together. So basically before the last refactoring regarding the deepLink all the regression tests succeeded. After that we added the tests but if you remember we also changed a bit the way sync-engine and UI communicate. IMHO I would re-run then cause many part are changed and having another check would be good.

typfel commented:
> @typfel we can decide together. So basically before the last refactoring regarding the deepLink all the regression tests succeeded. After that we added the tests but if you remember we also changed a bit the way sync-engine and UI communicate. IMHO I would re-run then cause many part are changed and having another check would be good.

Make sense to me, then let's re-run the tests and merge if no tests fail.

marcomaddy1982 commented:
> > @typfel we can decide together. So basically before the last refactoring regarding the deepLink all the regression tests succeeded. After that we added the tests but if you remember we also changed a bit the way sync-engine and UI communicate. IMHO I would re-run then cause many part are changed and having another check would be good.
> 
> Make sense to me, then let's re-run the tests and merge if no tests fail.

@typfel The playground build is made and QA will start later the regression tests.

David-Henner commented on WireDomain/Sources/WireDomain/Synchronization/PullPendingUpdateEventsV2Sync.swift:
```diff
@@ -134,7 +134,10 @@ public struct PullPendingUpdateEventsSyncV2: PullPendingUpdateEventsSyncV2Protoc
+        let backgroundTaskManager = coreCryptoProvider.backgroundTaskManager
+        try await coreCryptoProvider.coreCrypto().transaction(
+            backgroundTaskManager: backgroundTaskManager
+        ) { coreCryptoContext in
```
if the `backgroundTaskManager` and `coreCrypto` are  both held by `coreCryptoProvider`, is it necessary to pass the manager for every transaction ?

johnxnguyen replied:
The CoreCrypto provider doesn't perform the actual transaction so it doesn't know to pass the background task manager to core crypto. I know that it's not a very elegant solution and perhaps we can find a way to improve it. However I did discuss with @typfel that this perhaps could be moved inside core crypto if we can validate the solution on the iOS side first.

David-Henner replied:
Understood, thanks :) Perhaps you could add a TODO with a ticket to reflect this ?

David-Henner commented on wire-ios-data-model/Source/Core Crypto/CoreCryptoProvider.swift:
```diff
@@ -409,3 +427,44 @@ public actor CoreCryptoProvider: CoreCryptoProviderProtocol {
+                    "Background task expiring, cancelling core crypto transaction",
+                    attributes: .safePublic
+                )
+                transactionTask.cancel()
```
so, cancelling the task ensures that core crypto releases the lock?

johnxnguyen replied:
This would depend on cancellation checks in the given `block`, so no it does not guarantee the transaction is cancelled. I did not check extensively each transaction so that is something to do, but I would also be surprised if there is a transaction that lasts long enough for the task to expire.

netbe commented on WireFoundation/Sources/WireUtilitiesPackage/FileArchiving/BackgroundTaskManager.swift:
```diff
@@ -0,0 +1,74 @@
+/// Since the BackgroundTaskManager is only intended to be
+/// used from a main app target, this implementation can be used
+/// when running in an app extension.
+public struct NoOpBackgroundTaskManager: BackgroundTaskManager {
```
question: so we expect that NSE and ShareExtension never hold a file lock when suspended ?

johnxnguyen replied:
In the case of the NSE, either it completes or it expires. In the case that it expires we cancel the ongoing task and it should clean up gracefully. For the share extension I didn't look yet so I'm not sure, do you think it could be a problem?

netbe replied:
the usage is on foreground, I haven't seen any crash on share ext, so ok for now

netbe reviewed (commented):
left a question

netbe commented on wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift:
```diff
@@ -1653,4 +1653,19 @@ extension ZMUserSession {
+public extension ZMUserSession {
+
+    // TODO: [WPB-25551] Clean up testing code
+    func _simulateLongCCTransaction(seconds: Int) async throws {
```
praise: good to have a way to test this

question: did you try this with the previous implementation? with the ProcessInfo imp

johnxnguyen replied:
Yes I did and it also was useful

netbe reviewed (approved):
Looks good, however it's not completely clear to me if the approach with ProcessInfo was not working @johnxnguyen

netbe commented on wire-ios-data-model/Source/Core Crypto/CoreCryptoProvider.swift:
```diff
@@ -428,7 +418,37 @@ public actor CoreCryptoProvider: CoreCryptoProviderProtocol {
-public extension CoreCryptoProtocol {
+/// A wrapper object for CoreCrypto that ensures transactions are executed
+/// within a background task (if a task manager is provided).
+public final class SafeCoreCrypto {
```
suggestion: move to own file

johnxnguyen replied:
done
